### PR TITLE
Support broker user

### PR DESCRIPTION
**Pull request by omercs**
Added getBrokerUsers method to return UserInfo from Authenticator side. Account obj has only username. It is not providing userid.

Fix: Additional user fields are stored in the user data. ADAL access this through authenticator API and returns the UserInfo array.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23issuecomment-70689316%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23issuecomment-70775940%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245236%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245385%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245471%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245572%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23246000%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23246321%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23269361%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23269407%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23270140%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23270280%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23issuecomment-70689316%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Ping%22%2C%20%22created_at%22%3A%20%222015-01-20T16%3A56%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/466805%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/omercs%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222015-01-21T02%3A49%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/afshins%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%208f11caff88045b1f884db70b68c588c79134b268%20src/src/com/microsoft/aad/adal/AuthenticationActivity.java%20210%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245236%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Call%20result.taskResult.getUserInfo%28%29%20once%20and%20save%20it%20in%20a%20variable%22%2C%20%22created_at%22%3A%20%222015-01-20T18%3A54%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/afshins%22%7D%7D%2C%20%7B%22body%22%3A%20%22Good%20suggestion.%22%2C%20%22created_at%22%3A%20%222015-01-21T00%3A23%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/466805%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/omercs%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/src/com/microsoft/aad/adal/AuthenticationActivity.java%3AL995-1023%22%7D%2C%20%22Pull%208f11caff88045b1f884db70b68c588c79134b268%20src/src/com/microsoft/aad/adal/AuthenticationContext.java%2032%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245471%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Does%20Java%20have%20%3F%20operator%3F%22%2C%20%22created_at%22%3A%20%222015-01-20T18%3A57%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/afshins%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/src/com/microsoft/aad/adal/AuthenticationContext.java%3AL268-291%22%7D%2C%20%22Pull%208f11caff88045b1f884db70b68c588c79134b268%20src/src/com/microsoft/aad/adal/BrokerProxy.java%2094%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23246321%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20name%20of%20this%20method%20is%20a%20bit%20strange.%20Are%20you%20updating%20credentials%20as%20part%20of%20a%20request%20to%20get%20broker%20users%3F%22%2C%20%22created_at%22%3A%20%222015-01-20T19%3A07%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/afshins%22%7D%7D%2C%20%7B%22body%22%3A%20%22There%20are%20limited%20api%20methods%20on%20AccountManager%20plugin.%20Adal%20uses%20this%20method%20to%20pass%20a%20flag%20as%20DATA_USER_INFO%20with%20bundle%20options%20to%20execute%20methods%20inside%20the%20authenticator%20plugin.%20AccountManager%20API%20is%20fixed.%22%2C%20%22created_at%22%3A%20%222015-01-21T00%3A41%3A35Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/466805%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/omercs%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/src/com/microsoft/aad/adal/BrokerProxy.java%3AL457-511%22%7D%2C%20%22Pull%208f11caff88045b1f884db70b68c588c79134b268%20src/src/com/microsoft/aad/adal/BrokerProxy.java%2079%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23246000%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Add%20a%20comment%20why%20this%20is%20a%20problem.%22%2C%20%22created_at%22%3A%20%222015-01-20T19%3A02%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/afshins%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/src/com/microsoft/aad/adal/BrokerProxy.java%3AL457-511%22%7D%2C%20%22Pull%208f11caff88045b1f884db70b68c588c79134b268%20src/src/com/microsoft/aad/adal/AuthenticationActivity.java%20242%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245385%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Adding%20logs%20in%20general%20is%20good%2C%20but%20it%20is%20not%20a%20good%20idea%20to%20add%20expensive%20calls%20just%20to%20add%20log.%20Could%20this%20be%20such%20a%20case%3F%22%2C%20%22created_at%22%3A%20%222015-01-20T18%3A56%3A20Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/afshins%22%7D%7D%2C%20%7B%22body%22%3A%20%22This%20is%20important%20to%20track%20and%20related%20to%20the%20issue.%22%2C%20%22created_at%22%3A%20%222015-01-21T00%3A23%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/466805%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/omercs%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/src/com/microsoft/aad/adal/AuthenticationActivity.java%3AL1024-1039%22%7D%2C%20%22Pull%208f11caff88045b1f884db70b68c588c79134b268%20src/src/com/microsoft/aad/adal/BrokerProxy.java%2019%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-android/pull/338%23discussion_r23245572%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20is%20a%20complex%20logic.%20Add%20some%20comment.%22%2C%20%22created_at%22%3A%20%222015-01-20T18%3A58%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/afshins%22%7D%7D%2C%20%7B%22body%22%3A%20%22Added%22%2C%20%22created_at%22%3A%20%222015-01-21T00%3A38%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/466805%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/omercs%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/src/com/microsoft/aad/adal/BrokerProxy.java%3AL89-119%22%7D%7D%2C%20%22approved%22%3A%20%7B%22https%3A//github.com/afshins%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3790715%3Fv%3D3%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/afshins'></a>

- [ ] <a href='#crh-comment-Pull 8f11caff88045b1f884db70b68c588c79134b268 src/src/com/microsoft/aad/adal/AuthenticationActivity.java 210'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338#discussion_r23245236'>File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java:L995-1023</a></b>
- <a href='https://github.com/afshins'></a> Call result.taskResult.getUserInfo() once and save it in a variable
- <a href='https://github.com/omercs'></a> Good suggestion.
- [ ] <a href='#crh-comment-Pull 8f11caff88045b1f884db70b68c588c79134b268 src/src/com/microsoft/aad/adal/AuthenticationActivity.java 242'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338#discussion_r23245385'>File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java:L1024-1039</a></b>
- <a href='https://github.com/afshins'></a> Adding logs in general is good, but it is not a good idea to add expensive calls just to add log. Could this be such a case?
- <a href='https://github.com/omercs'></a> This is important to track and related to the issue.
- [ ] <a href='#crh-comment-Pull 8f11caff88045b1f884db70b68c588c79134b268 src/src/com/microsoft/aad/adal/AuthenticationContext.java 32'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338#discussion_r23245471'>File: src/src/com/microsoft/aad/adal/AuthenticationContext.java:L268-291</a></b>
- <a href='https://github.com/afshins'></a> Does Java have ? operator?
- [ ] <a href='#crh-comment-Pull 8f11caff88045b1f884db70b68c588c79134b268 src/src/com/microsoft/aad/adal/BrokerProxy.java 19'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338#discussion_r23245572'>File: src/src/com/microsoft/aad/adal/BrokerProxy.java:L89-119</a></b>
- <a href='https://github.com/afshins'></a> This is a complex logic. Add some comment.
- <a href='https://github.com/omercs'></a> Added
- [ ] <a href='#crh-comment-Pull 8f11caff88045b1f884db70b68c588c79134b268 src/src/com/microsoft/aad/adal/BrokerProxy.java 79'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338#discussion_r23246000'>File: src/src/com/microsoft/aad/adal/BrokerProxy.java:L457-511</a></b>
- <a href='https://github.com/afshins'></a> Add a comment why this is a problem.
- [ ] <a href='#crh-comment-Pull 8f11caff88045b1f884db70b68c588c79134b268 src/src/com/microsoft/aad/adal/BrokerProxy.java 94'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338#discussion_r23246321'>File: src/src/com/microsoft/aad/adal/BrokerProxy.java:L457-511</a></b>
- <a href='https://github.com/afshins'></a> The name of this method is a bit strange. Are you updating credentials as part of a request to get broker users?
- <a href='https://github.com/omercs'></a> There are limited api methods on AccountManager plugin. Adal uses this method to pass a flag as DATA_USER_INFO with bundle options to execute methods inside the authenticator plugin. AccountManager API is fixed.
- [x] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338#issuecomment-70689316'>General Comment</a></b>
- <a href='https://github.com/omercs'></a> Ping

<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-library-for-android/pull/338?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-library-for-android/pull/338?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-library-for-android/pull/338?approve=0'></a>&nbsp;<a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by RandalliLama** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/338'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by omercs** [comment]
Ping

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 1]
```diff
@@ -985,15 +995,28 @@ private void setAccount(final TokenTaskResult result) throws InvalidKeyException
...
+                    Logger.i(TAG, "Set userinfo from account", "");
                     result.taskResult.setUserInfo(new UserInfo(name, name, "", "", name));
                     mRequest.setLoginHint(name);
                 }
-
+                else
+                {
+                    Logger.i(TAG, "Saving userinfo to account", "");
+                    mAccountManager.setUserData(newaccount, AuthenticationConstants.Broker.ACCOUNT_USERINFO_USERID,
+                            result.taskResult.getUserInfo().getUserId());
```
Comment (afshins) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: Call result.taskResult.getUserInfo() once and save it in a variable

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 2]
```diff
@@ -1001,10 +1024,15 @@ private void setAccount(final TokenTaskResult result) throws InvalidKeyException
...
                 Gson gson = new Gson();
-                Logger.v(TAG, "app context:" + getApplicationContext().getPackageName()
+                Logger.i(TAG, "app context:" + getApplicationContext().getPackageName()
                         + " context:" + AuthenticationActivity.this.getPackageName()
-                        + " calling packagename:" + getCallingPackage());
+                        + " calling packagename:" + getCallingPackage(), "");
                 StorageHelper cryptoHelper = new StorageHelper(getApplicationContext());
+
+                if (AuthenticationSettings.INSTANCE.getSecretKeyData() == null) {
+                    Logger.i(TAG, "setAccount: user key is null", "");
```
Comment (afshins) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: Adding logs in general is good, but it is not a good idea to add expensive calls just to add log. Could this be such a case?

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 3]
```diff
@@ -265,6 +268,23 @@ public String getBrokerUser() {
...
+     * @return user {@link UserInfo}
+     * 
+     * @throws IOException
+     * @throws AuthenticatorException
+     * @throws OperationCanceledException
+     */
+    public UserInfo[] getBrokerUsers() throws OperationCanceledException, AuthenticatorException,
+            IOException {
+        if (mBrokerProxy != null) {
+            return mBrokerProxy.getBrokerUsers();
```
Comment (afshins) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: Does Java have ? operator?

File: src/src/com/microsoft/aad/adal/BrokerProxy.java [hunk 4]
```diff
@@ -87,28 +89,30 @@ public BrokerProxy(final Context ctx) {
     @Override
     public boolean canSwitchToBroker() {
         String packageName = mContext.getPackageName();
-        return !AuthenticationSettings.INSTANCE.getSkipBroker() && verifyManifestPermissions()
-                && !packageName.equalsIgnoreCase(AuthenticationSettings.INSTANCE.getBrokerPackageName())
-                && verifyAuthenticator(mAcctManager) && verifyAccount();
+        return !AuthenticationSettings.INSTANCE.getSkipBroker()
+                && verifyManifestPermissions()
+                && !packageName.equalsIgnoreCase(AuthenticationSettings.INSTANCE
+                        .getBrokerPackageName()) && verifyAuthenticator(mAcctManager)
```
Comment (afshins) [inline src/src/com/microsoft/aad/adal/BrokerProxy.java]: This is a complex logic. Add some comment.

File: src/src/com/microsoft/aad/adal/BrokerProxy.java [hunk 5]
```diff
@@ -452,4 +457,54 @@ private boolean verifyAuthenticator(final AccountManager am) {
...
+     * @throws IOException
+     * @throws AuthenticatorException
+     * @throws OperationCanceledException
+     */
+    @Override
+    public UserInfo[] getBrokerUsers() throws OperationCanceledException, AuthenticatorException,
+            IOException {
+
+        if (Looper.myLooper() == Looper.getMainLooper()) {
+            throw new IllegalArgumentException("Calling getBrokerUsers on main thread");
```
Comment (afshins) [inline src/src/com/microsoft/aad/adal/BrokerProxy.java]: Add a comment why this is a problem.

File: src/src/com/microsoft/aad/adal/BrokerProxy.java [hunk 6]
```diff
@@ -452,4 +457,54 @@ private boolean verifyAuthenticator(final AccountManager am) {
...
+        bundle.putBoolean(DATA_USER_INFO, true);
+
+        if (accountList != null) {
+
+            // get info for each user
+            UserInfo[] users = new UserInfo[accountList.length];
+            for (int i = 0; i < accountList.length; i++) {
+                AccountManagerFuture<Bundle> result = mAcctManager.updateCredentials(
+                        accountList[i], AuthenticationConstants.Broker.AUTHTOKEN_TYPE, bundle,
+                        null, null, null);
```
Comment (afshins) [inline src/src/com/microsoft/aad/adal/BrokerProxy.java]: The name of this method is a bit strange. Are you updating credentials as part of a request to get broker users?

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java (hunk 1 above)
Comment (omercs) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: Good suggestion.

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java (hunk 2 above)
Comment (omercs) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: This is important to track and related to the issue.

File: src/src/com/microsoft/aad/adal/BrokerProxy.java (hunk 4 above)
Comment (omercs) [inline src/src/com/microsoft/aad/adal/BrokerProxy.java]: Added

File: src/src/com/microsoft/aad/adal/BrokerProxy.java (hunk 6 above)
Comment (omercs) [inline src/src/com/microsoft/aad/adal/BrokerProxy.java]: There are limited api methods on AccountManager plugin. Adal uses this method to pass a flag as DATA_USER_INFO with bundle options to execute methods inside the authenticator plugin. AccountManager API is fixed.

**Comment by afshins** [comment]
:shipit:

**Pull request merged**